### PR TITLE
Examples

### PR DESCRIPTION
**Pull request by pb82**
Adds examples for providing external admin credentials and also updates the code to lookup the credentials so that the Operator can discover it.

Contains the following examples:

* Basic
* Credentials from config
* Credentials from secret
* OAuth Proxy

ping @HubertStefanski @NissesSenap @weisdd

### CODE REVIEW

File: api/v1beta1/typeoverrides.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ type DeploymentV1Spec struct {
 
 	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
 
-	Template *v14.PodTemplateSpec `json:"template,omitempty" protobuf:"bytes,3,opt,name=template"`
+	Template *DeploymentV1PodTemplateSpec `json:"template,omitempty" protobuf:"bytes,3,opt,name=template"`
```
Comment (pb82) [inline api/v1beta1/typeoverrides.go:61]: This was required so we can override certain fields of the pod template without providing others.

File: examples/basic/resources.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        "list": []
+      },
+      "annotations": {
+        "list": []
+      },
+      "refresh": "5s",
+      "schemaVersion": 17,
+      "version": 0,
+      "links": []
+    }
```
Comment (weisdd) [inline examples/basic/resources.yaml:54]: README.md says: "A basic deployment of Grafana with a dashboard and a datasource.", but the datasource is missing. So, either README.md or the example should be fixed :)

**Review by weisdd — CHANGES_REQUESTED** [review]

File: controllers/client/grafana_client.go [hunk 3]
```diff
@@ -2,51 +2,97 @@ package client
...
+			Namespace: grafana.Namespace,
+		}
+		err := c.Get(ctx, selector, secret)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if secret.Data == nil {
+			return nil, nil
```
[the remark this answers is not in the document]
Comment (pb82) [reply controllers/client/grafana_client.go:49]: Yes, good point

File: controllers/metrics/metrics.go [hunk 4]
```diff
@@ -26,10 +26,18 @@ var (
 		Name:      "requests",
 		Help:      "requests against the grafana api per instance",
 	}, []string{"instance_name", "path", "method", "status"})
+
+	DashboardUrlRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
```
Comment (weisdd) [inline controllers/metrics/metrics.go:30]: Example of such metric: `grafana_operator_dashboards_requests{dashboard="grafanadashboard-from-url",method="GET",path="/integr8ly/grafana-operator/master/deploy/examples/remote/grafana-dashboard.json",status="200"} 2`
As you can see, there's no reference for a namespace, so I wonder what would happen when you have dashboards with the same name in multiple namespaces?

Comment (pb82) [reply controllers/metrics/metrics.go:30]: Good point. I wonder, do we even need to include the dashboard? The metric is meant to count the requests against the dashboard API overall, it doesn't have to be per dashboard.

Comment (weisdd) [reply controllers/metrics/metrics.go:30]: I think we're talking about different metrics here. My point was only about counting the number of requests for dashboards pulled from URLs. :)

Comment (pb82) [reply controllers/metrics/metrics.go:30]: Ah I see. This metric does count the number of requests against the Grafana dashboard API. I've now added the namespace to the dashboard name label.

**Comment by pb82** [comment]
@weisdd are you ok with merging this?

**Comment by weisdd** [comment]
@pb82 yep, let's merge it, and thanks for the fixes :)

**Pull request merged**